cmake/toolchain.cmake

```diff
@@ -0,0 +1,51 @@
+# SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+set(CMAKE_C_FLAGS_RELEASE_INIT "-O3")
+set(CMAKE_CXX_FLAGS_RELEASE_INIT "-O3")
+
+set(CMAKE_C_FLAGS_RELWITHDEBINFO_INIT "-O3")
+set(CMAKE_CXX_FLAGS_RELWITHDEBINFO_INIT "-O3")
+
+# BEMAN_BUILDSYS_SANITIZER is not a general use option
+# It is used by preset and CI system.
+# There's three possible values:
+# TSan: Thread sanitizer
+# ASan: All sanitizer (majorly Address sanitizer) that doesn't conflict with TSan
+# OFF: No sanitizer
+
+if(DEFINED BEMAN_BUILDSYS_SANITIZER)
+    if(BEMAN_BUILDSYS_SANITIZER STREQUAL "ASan")
+        # Basic ASan flags
+        set(SANITIZER_FLAGS
+            "-fsanitize=address -fsanitize=pointer-compare -fsanitize=pointer-subtract -fsanitize=undefined"
+        )
+
+        # Exclude -fsanitize=leak if using GCC on macOS
+        # TODO: Is there a way to detect Apple Clang???
+        if(
+            CMAKE_CXX_COMPILER_ID STREQUAL "GNU"
+            AND CMAKE_SYSTEM_NAME STREQUAL "Darwin"
+        )
+            message(STATUS "Using GCC on macOS; excluding -fsanitize=leak")
+        else()
+            set(SANITIZER_FLAGS "${SANITIZER_FLAGS} -fsanitize=leak")
+        endif()
```

Comment:
Is there a better way to detect Apple Clang faking gcc?

Reply:
I would expect that the compiler identification should figure out if it's really AppleClang even though /usr/bin/g++ is symlinked to clang? 
https://cmake.org/cmake/help/latest/variable/CMAKE_LANG_COMPILER_ID.html

Reply:
I don't have a macos development environment to test this.

Is AppleClang the identifier?

Reply:
@steve-downey steve can you explicitly block this PR in review so someone doesn't accidentally come in and merge this?

Recommends message:
We need to verify `CMAKE_CXX_COMPILER_ID` for g++ on macos is `AppleClang`.

Reply:
```bash
iMac:exemplar clausklein$ CXX=g++-14 cmake -S . -B build -G Ninja
-- The CXX compiler identification is GNU 14.2.0
-- Checking whether CXX compiler has -isysroot
-- Checking whether CXX compiler has -isysroot - yes
-- Checking whether CXX compiler supports OSX deployment target flag
-- Checking whether CXX compiler supports OSX deployment target flag - yes
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /usr/local/bin/g++-14 - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- The C compiler identification is AppleClang 16.0.0.16000026
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/cc - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Found Python3: /usr/local/Frameworks/Python.framework/Versions/3.13/bin/python3.13 (found version "3.13.0") found components: Interpreter
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD - Success
-- Found Threads: TRUE
-- Looking for __cpp_lib_ranges
-- Looking for __cpp_lib_ranges - not found
CMake Warning at examples/CMakeLists.txt:12 (message):
  Missing range support! Skip: identity_as_default_projection


Examples to be built: identity_direct_usage
-- Configuring done (16.8s)
-- Generating done (0.1s)
-- Build files have been written to: /Users/clausklein/Workspace/cpp/beman-project/exemplar/build
iMac:exemplar clausklein$ 
```

Reply:
Ah! Is the compiler id for AppleClang GNU?!

Reply:
I had to changed this:
```diff
diff --git a/cmake/toolchain.cmake b/cmake/toolchain.cmake
index b3cd128..450e8a9 100644
--- a/cmake/toolchain.cmake
+++ b/cmake/toolchain.cmake
@@ -21,10 +21,13 @@ if(DEFINED BEMAN_BUILDSYS_SANITIZER)
         )
 
         # Exclude -fsanitize=leak on Apple Clang (gcc on macos) as it is not supported.
-        if(CMAKE_CXX_COMPILER_ID STREQUAL "AppleClang")
+        if(NOT CMAKE_CXX_COMPILER_ID)
+            message(WARNING "toolchain is used before CMAKE_CXX_COMPILER_ID was set!")
+        endif()
+        if(APPLE) # AND ((${CMAKE_CXX_COMPILER_ID} MATCHES "Clang") OR (${CMAKE_CXX_COMPILER_ID} MATCHES "GNU")))
             message(STATUS "Using GCC on macOS; excluding -fsanitize=leak")
         else()
-            set(SANITIZER_FLAGS "${SANITIZER_FLAGS} -fsanitize=leak")
+            list(APPEND SANITIZER_FLAGS "-fsanitize=leak")
         endif()
     elseif(BEMAN_BUILDSYS_SANITIZER STREQUAL "TSan")
         # Basic ASan flags
@@ -38,10 +41,6 @@ if(DEFINED BEMAN_BUILDSYS_SANITIZER)
         )
     endif()
 
-    set(CMAKE_C_FLAGS_DEBUG_INIT
-        "${CMAKE_C_FLAGS_DEBUG_INIT} ${SANITIZER_FLAGS}"
-    )
-    set(CMAKE_CXX_FLAGS_DEBUG_INIT
-        "${CMAKE_C_FLAGS_DEBUG_INIT} ${SANITIZER_FLAGS}"
-    )
+    list(APPEND CMAKE_C_FLAGS_DEBUG_INIT "${SANITIZER_FLAGS}")
+    list(APPEND CMAKE_CXX_FLAGS_DEBUG_INIT "${SANITIZER_FLAGS}")
 endif()
```

Reply:
And the string, list handling in cmake is petty!
```bash
ninja -C build/gcc-debug -j1
ninja: Entering directory `build/gcc-debug'
[1/12] Building CXX object _deps/googletest-build/googletest/CMakeFiles/gtest.dir/src/gtest-all.cc.o
FAILED: _deps/googletest-build/googletest/CMakeFiles/gtest.dir/src/gtest-all.cc.o 
/usr/bin/g++  -I/Users/clausklein/Workspace/cpp/beman-project/exemplar/build/gcc-debug/_deps/googletest-src/googletest/include -I/Users/clausklein/Workspace/cpp/beman-project/exemplar/build/gcc-debug/_deps/googletest-src/googletest -fsanitize=address -fsanitize=pointer-compare -fsanitize=pointer-subtract -fsanitize=undefined;-fsanitize=address -fsanitize=pointer-compare -fsanitize=pointer-subtract -fsanitize=undefined -g -std=c++20 -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX15.1.sdk -mmacosx-version-min=14.7  -DGTEST_HAS_PTHREAD=1 -MD -MT _deps/googletest-build/googletest/CMakeFiles/gtest.dir/src/gtest-all.cc.o -MF _deps/googletest-build/googletest/CMakeFiles/gtest.dir/src/gtest-all.cc.o.d -o _deps/googletest-build/googletest/CMakeFiles/gtest.dir/src/gtest-all.cc.o -c /Users/clausklein/Workspace/cpp/beman-project/exemplar/build/gcc-debug/_deps/googletest-src/googletest/src/gtest-all.cc
clang++: error: no input files
/bin/sh: -fsanitize=address: command not found
ninja: build stopped: subcommand failed.
```

Reply:
wait why is compiler identification GNU here but AppleClang below?